### ISSUE TITLE
BinaryView Memory Leak in DebuggerController When a ReportCollection Tab is Closed Last

### ISSUE THREAD

**Issue by bpotchik** [Vector35/debugger#641]
0. open file
1. open report widget (current_function.request_debug_report("mlil"))
2. close the tab/file for that view
3. close the report widget tab
4. get_memory_usage_info()

**Comment by xusheng6** [comment]
This is how the relevant events unfold:

1. when we close the tab for the view itself, the debugger actually correctly detects the closing of the last tab that belongs to the view, and destroys the controller for it
2. The report widget automatically becomes the current tab
3. The report widget has a reference to the binary view, and SetCurrentBinaryView is called with it
4. The Python console receives it, and constructs a new DebuggerController using that view
5. When the report widget is closed, `NotificationListener::OnBeforeCloseFile` is not called, so the debugger does not get a chance to know about
6. Memory leak

I do not know what is a proper fix for this because:
1. The main windows seems to intentionally NOT call `OnBeforeCloseFile` for widget like report widget, settings widget. Which I think is a wise choice
2. One might argue that the closing of the view tab is NOT technically the last tab of the view (the report widget is still open and it has a reference to the view). But that does not help, since even if we do not destroy the old controller when that happens, we still would not be able to know the reprot tab is finally being closed at all. 

To summarize, the debugger relies on `OnBeforeCloseFile` to know when a file is closed and then cleanup the controller for it. Now the report widget is the last tab that gets closed, and `OnBeforeCloseFile` is NOT called for it. 

I also believe that this also affects the database saving logic in the very same way, since the way the debugger detects the "last open tab" of the file is the same as how main window does it. If we modify the view, and then open the report widget, we will be asked to save the database when the view tab is closed. However, you can still make changes to the view using the Python console, and the main window will not ask you to save it

**Comment by bpotchik** [comment]
One option to solve this is to parent any report widgets to the view tab. When the last view tab is closed, all other report tabs with that view are closed first.

**Comment by xusheng6** [comment]
I am bumping this since this is not really a regression -- it has been there for a while and we happen to discover it recently
